source/opt/ir_builder.h

```diff
@@ -0,0 +1,144 @@
+                "Only Def-use analysis update is supported");
+
+ public:
+  using InsertionPointTy = spvtools::ir::BasicBlock::iterator;
+
+  InstructionBuilder(ir::IRContext* context, InsertionPointTy insert_before)
+      : context_(context), insert_before_(insert_before) {}
+
+  // Creates a new selection merge instruction.
+  // The id |merge_id| is the merge basic block id.
+  ir::Instruction* AddSelectionMerge(uint32_t merge_id) {
+    std::unique_ptr<ir::Instruction> new_branch_merge(new ir::Instruction(
+        GetContext(), SpvOpSelectionMerge, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {merge_id}},
+         {spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER, {0}}}));
+    return AddInstruction(std::move(new_branch_merge));
+  }
+
+  // Creates a new branch instruction to |label_id|.
+  // Note that the user must makes sure the final basic block is
+  // well formed.
+  ir::Instruction* AddBranch(uint32_t label_id) {
+    std::unique_ptr<ir::Instruction> new_branch(new ir::Instruction(
+        GetContext(), SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {label_id}}}));
+    return AddInstruction(std::move(new_branch));
+  }
+
+  // Creates a new conditional instruction and the associated selection merge
+  // instruction if requested.
+  // The id |cond_id| is the id to the the condition instruction, must be of
+  // type bool.
+  // The id |true_id| is the basic block id to branch to if the condition is
+  // true.
+  // The id |false_id| is the basic block id to branch to if the condition is
+  // false.
+  // The id |merge_id| is the merge basic block id for the selection merge
+  // instruction. If |merge_id| equals kInvalidId then no selection merge
+  // instruction will be created.
+  // Note that the user must makes sure the final basic block is
+  // well formed.
+  ir::Instruction* AddBranchCond(uint32_t cond_id, uint32_t true_id,
+                                 uint32_t false_id,
+                                 uint32_t merge_id = kInvalidId) {
+    if (merge_id != kInvalidId) {
+      AddSelectionMerge(merge_id);
+    }
+    std::unique_ptr<ir::Instruction> new_branch(new ir::Instruction(
+        GetContext(), SpvOpBranchConditional, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {cond_id}},
+         {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {true_id}},
+         {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {false_id}}}));
+    return AddInstruction(std::move(new_branch));
+  }
+
+  // Creates a phi instruction.
+  // The id |type| must be the id of the phi instruction's type.
+  // The vector |incomings| must be a sequence of pairs of <definition id,
+  // parent id>.
+  ir::Instruction* AddPhi(uint32_t type,
+                          const std::vector<uint32_t>& incomings) {
+    assert(incomings.size() % 2 == 0 && "A sequence of pairs is expected");
+    std::vector<ir::Operand> phi_ops;
+    for (size_t i = 0; i < incomings.size(); i += 2) {
```

Comment:
I get your thinking here, but I feel like it would be better to go with the simpler code of adding each input operand separately (e.g. increment by 1 and add 1 element in the body).

Reply:
I understand, on the other hand, I always found disturbing when you get an extra increment inside the body of a for loop.

In the code base, phi node handling are commonly done using this idiom. Is that legacy you are trying to get rid of ?

Reply:
Not per se, but in this case I think it is unnecessary. You are simple iterating through each input. While they are paired logically, I think the code benefits only using the idiom when necessary.